File: intercom_flutter/android/src/main/kotlin/io/maido/intercom/IntercomFlutterPlugin.kt
package io.maido.intercom

import android.app.Application
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.intercom.android.sdk.*
import io.intercom.android.sdk.identity.Registration
import io.intercom.android.sdk.push.IntercomPushClient

class IntercomFlutterPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
  companion object {
    @JvmStatic
    lateinit var application: Application

    @JvmStatic
    fun initSdk(application: Application, appId: String, androidApiKey: String) {
      Intercom.initialize(application, apiKey = androidApiKey, appId = appId)
    }
  }

  private val intercomPushClient = IntercomPushClient()
  private var unreadConversationCountListener: UnreadConversationCountListener? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding.binaryMessenger, "maido.io/intercom")
    channel.setMethodCallHandler(IntercomFlutterPlugin())
    val unreadEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "maido.io/intercom/unread")
    unreadEventChannel.setStreamHandler(IntercomFlutterPlugin())
    application = flutterPluginBinding.applicationContext as Application
  }

  // https://stackoverflow.com/a/62206235
  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    application = binding.activity.application
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "initialize" -> {
        val apiKey = call.argument<String>("androidApiKey")
        val appId = call.argument<String>("appId")
        Intercom.initialize(application, apiKey, appId)
        result.success("Intercom initialized")
      }
      "setBottomPadding" -> {
        val padding = call.argument<Int>("bottomPadding")
        if (padding != null) {
          Intercom.client().setBottomPadding(padding)
          result.success("Bottom padding set")
        }
      }
      "setUserHash" -> {
        val userHash = call.argument<String>("userHash")
        if (userHash != null) {
          Intercom.client().setUserHash(userHash)
          result.success("User hash added")
        }
      }
      "loginIdentifiedUserWithUserId" -> {
        val userId = call.argument<String>("userId")
        if (userId != null) {
          var registration = Registration.create()
          registration = registration.withUserId(userId)
          Intercom.client().loginIdentifiedUser(registration, intercomStatusCallback = object : IntercomStatusCallback {
            override fun onFailure(intercomError: IntercomError) {
              // Handle failure
              result.error(intercomError.errorCode.toString(), intercomError.errorMessage, getIntercomError(
                  errorCode = intercomError.errorCode,
                  errorMessage = intercomError.errorMessage,
              ))
            }

            override fun onSuccess() {
              // Handle success
              result.success("User created")
            }
          })
        }
      }
      "loginIdentifiedUserWithEmail" -> {
        val email = call.argument<String>("email")
        if (email != null) {
          var registration = Registration.create()
          registration = registration.withEmail(email)
          Intercom.client().loginIdentifiedUser(registration, intercomStatusCallback = object : IntercomStatusCallback {
            override fun onFailure(intercomError: IntercomError) {
              // Handle failure
              result.error(intercomError.errorCode.toString(), intercomError.errorMessage, getIntercomError(
                  errorCode = intercomError.errorCode,
                  errorMessage = intercomError.errorMessage,
              ))
            }

            override fun onSuccess() {
              // Handle success
              result.success("User created")
            }
          })
        }
      }
      "loginUnidentifiedUser" -> {
        Intercom.client().loginUnidentifiedUser(intercomStatusCallback = object : IntercomStatusCallback {
          override fun onFailure(intercomError: IntercomError) {
            // Handle failure
            result.error(intercomError.errorCode.toString(), intercomError.errorMessage, getIntercomError(
                errorCode = intercomError.errorCode,
                errorMessage = intercomError.errorMessage,
            ))
          }

          override fun onSuccess() {
            // Handle success
            result.success("User created")
          }
        })
      }
      "logout" -> {
        Intercom.client().logout()
        result.success("logout")
      }
      "setLauncherVisibility" -> {
        val visibility = call.argument<String>("visibility")
        if (visibility != null) {
          Intercom.client().setLauncherVisibility(Intercom.Visibility.valueOf(visibility))
          result.success("Showing launcher: $visibility")
        }
      }
      "displayMessenger" -> {
        Intercom.client().present()
        result.success("Launched")
      }
      "hideMessenger" -> {
        Intercom.client().hideIntercom()
        result.success("Hidden")
      }
      "displayHelpCenter" -> {
        Intercom.client().present(IntercomSpace.HelpCenter)
        result.success("Launched")
      }
      "displayHelpCenterCollections" -> {
        val collectionIds = call.argument<ArrayList<String>>("collectionIds")
        Intercom.client().presentContent(
          content = IntercomContent.HelpCenterCollections(
            ids = collectionIds ?: emptyList()
          )
        )
        result.success("Launched")
      }
      "displayMessages" -> {
        Intercom.client().present(IntercomSpace.Messages)
        result.success("Launched")
      }
      "setInAppMessagesVisibility" -> {
        val visibility = call.argument<String>("visibility")
        if (visibility != null) {
          Intercom.client().setInAppMessageVisibility(Intercom.Visibility.valueOf(visibility))
          result.success("Showing in app messages: $visibility")
        } else {
          result.success("Launched")
        }
      }
      "unreadConversationCount" -> {
        val count = Intercom.client().unreadConversationCount
        result.success(count)
      }
      "updateUser" -> {
        Intercom.client().updateUser(getUserAttributes(call), intercomStatusCallback = object : IntercomStatusCallback {
          override fun onFailure(intercomError: IntercomError) {
            // Handle failure
            result.error(intercomError.errorCode.toString(), intercomError.errorMessage, getIntercomError(
                errorCode = intercomError.errorCode,
                errorMessage = intercomError.errorMessage,
            ))
          }

          override fun onSuccess() {
            // Handle success
            result.success("User updated")
          }
        })
      }
      "logEvent" -> {
        val name = call.argument<String>("name")
        val metaData = call.argument<Map<String, *>>("metaData")
        if (name != null) {
          Intercom.client().logEvent(name, metaData)
          result.success("Logged event")
        }
      }
      "sendTokenToIntercom" -> {
        val token = call.argument<String>("token")
        if (token != null) {
          intercomPushClient.sendTokenToIntercom(application, token)

          result.success("Token sent to Intercom")
        }
      }
      "handlePushMessage" -> {
        Intercom.client().handlePushMessage()
        result.success("Push message handled")
      }
      "displayMessageComposer" -> {
        if (call.hasArgument("message")) {
          Intercom.client().displayMessageComposer(call.argument("message"))
        } else {
          Intercom.client().displayMessageComposer()
        }
        result.success("Message composer displayed")
      }
      "isIntercomPush" -> {
        result.success(intercomPushClient.isIntercomPush(call.argument<Map<String, String>>("message")!!))
      }
      "handlePush" -> {
        intercomPushClient.handlePush(application, call.argument<Map<String, String>>("message")!!)
        result.success(null)
      }
      "displayArticle" -> {
        val articleId = call.argument<String>("articleId")
        if (articleId != null) {
          Intercom.client().presentContent(IntercomContent.Article(articleId))
          result.success("displaying article $articleId")
        }
      }
      "displayCarousel" -> {
        val carouselId = call.argument<String>("carouselId")
        if (carouselId != null) {
          Intercom.client().presentContent(IntercomContent.Carousel(carouselId))
          result.success("displaying carousel $carouselId")
        }
      }
      "displaySurvey" -> {
        val surveyId = call.argument<String>("surveyId")
        if (surveyId != null) {
          Intercom.client().presentContent(IntercomContent.Survey(surveyId))
          result.success("displaying survey $surveyId")
        }
      }
      "displayConversation" -> {
        val conversationId = call.argument<String>("conversationId")
        if (conversationId != null) {
          Intercom.client().presentContent(IntercomContent.Conversation(conversationId))
          result.success("displaying conversation $conversationId")
        }
      }
      "displayTickets" -> {
        Intercom.client().present(IntercomSpace.Tickets)
        result.success("Launched Tickets space")
      }
      "displayHome" -> {
        Intercom.client().present(IntercomSpace.Home)
        result.success("Launched Home space")
      }
      "isUserLoggedIn" -> {
        result.success(Intercom.client().isUserLoggedIn())
      }
      "fetchLoggedInUserAttributes" -> {
        val reg = Intercom.client().fetchLoggedInUserAttributes()
        val map = reg?.attributes?.toMap() ?: mutableMapOf<String, Any>()
        if(reg != null){
          // put the user_id and email from registration
          map["user_id"] = reg.userId
          map["email"] = reg.email
        }
        result.success(map)
      }
      else -> result.notImplemented()
    }
  }

  // generate a errorDetails object to pass
  private fun getIntercomError(errorCode: Int = -1, errorMessage: String = ""): Map<String, *> {
    return mapOf(
        "errorCode" to errorCode,
        "errorMessage" to errorMessage,
    )
  }

  // generate the user attributes
  private fun getUserAttributes(call: MethodCall): UserAttributes {
    // user attributes
    val name = call.argument<String>("name")
    val email = call.argument<String>("email")
    val phone = call.argument<String>("phone")
    val userId = call.argument<String>("userId")
    val company = call.argument<String>("company")
    val companyId = call.argument<String>("companyId")
    val customAttributes = call.argument<Map<String, Any?>>("customAttributes")
    val signedUpAt = call.argument<Any?>("signedUpAt")
    val language = call.argument<String>("language")

    val userAttributes = UserAttributes.Builder()

    if (name != null) {
      userAttributes.withName(name)
    }

    if (email != null) {
      userAttributes.withEmail(email)
    }

    if (phone != null) {
      userAttributes.withPhone(phone)
    }

    if (userId != null) {
      userAttributes.withUserId(userId)
    }

    if (company != null && companyId != null) {
      val icmCompany = Company.Builder()
      icmCompany.withName(company)
      icmCompany.withCompanyId(companyId)
      userAttributes.withCompany(icmCompany.build())
    }

    if (customAttributes != null) {
      for ((key, value) in customAttributes) {
        userAttributes.withCustomAttribute(key, value)
      }
    }

    val seconds: Long? = signedUpAt?.toString()?.toLongOrNull()
    if (seconds != null)
      userAttributes.withSignedUpAt(seconds)

    if (language != null) {
      userAttributes.withLanguageOverride(language)
    }

    return userAttributes.build()
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    unreadConversationCountListener = UnreadConversationCountListener { count -> events?.success(count) }
        .also {
          Intercom.client().addUnreadConversationCountListener(it)
        }
  }

  override fun onCancel(arguments: Any?) {
    if (unreadConversationCountListener != null) {
      Intercom.client().removeUnreadConversationCountListener(unreadConversationCountListener)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    if (unreadConversationCountListener != null) {
      Intercom.client().removeUnreadConversationCountListener(unreadConversationCountListener)
    }
  }

  override fun onDetachedFromActivity() {
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
  }

  override fun onDetachedFromActivityForConfigChanges() {
  }
}

File: intercom_flutter/ios/Classes/IntercomFlutterPlugin.m
#import "IntercomFlutterPlugin.h"
#import <Intercom/Intercom.h>

id unread;

@implementation UnreadStreamHandler
- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    unread = [[NSNotificationCenter defaultCenter] addObserverForName:IntercomUnreadConversationCountDidChangeNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        NSNumber *myNum = @([Intercom unreadConversationCount]);
        eventSink(myNum);
    }];
  return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    [[NSNotificationCenter defaultCenter] removeObserver:unread];
  return nil;
}
@end

@implementation IntercomFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    IntercomFlutterPlugin* instance = [[IntercomFlutterPlugin alloc] init];
    FlutterMethodChannel* channel =
    [FlutterMethodChannel methodChannelWithName:@"maido.io/intercom"
                                binaryMessenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:channel];
    FlutterEventChannel* unreadChannel = [FlutterEventChannel eventChannelWithName:@"maido.io/intercom/unread"
    binaryMessenger:[registrar messenger]];
    UnreadStreamHandler* unreadStreamHandler =
        [[UnreadStreamHandler alloc] init];
    [unreadChannel setStreamHandler:unreadStreamHandler];
    
}

- (void) handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result{
    if([@"initialize" isEqualToString:call.method]) {
        NSString *iosApiKey = call.arguments[@"iosApiKey"];
        NSString *appId = call.arguments[@"appId"];
        [Intercom setApiKey:iosApiKey forAppId:appId];
        result(@"Initialized Intercom");
    }
    else if([@"loginUnidentifiedUser" isEqualToString:call.method]) {
        [Intercom loginUnidentifiedUserWithSuccess:^{
            // Handle success
            result(@"Registered unidentified user");
        } failure:^(NSError * _Nonnull error) {
            // Handle error
            NSInteger errorCode = error.code;
            NSString *errorMsg = error.localizedDescription;
            
            result([FlutterError errorWithCode:[@(errorCode) stringValue]
                                       message:errorMsg
                                       details: [self getIntercomError:errorCode:errorMsg]]);
        }];
    }
    else if([@"setBottomPadding" isEqualToString:call.method]) {
        NSNumber *value = call.arguments[@"bottomPadding"];
        if(value != (id)[NSNull null] && value != nil) {
            CGFloat padding = [value doubleValue];
            [Intercom setBottomPadding:padding];
            result(@"Set bottom padding");
        }
    }
    else if([@"setUserHash" isEqualToString:call.method]) {
        NSString *userHash = call.arguments[@"userHash"];
        [Intercom setUserHash:userHash];
        result(@"User hash added");
    }
    else if([@"loginIdentifiedUserWithUserId" isEqualToString:call.method]) {
        NSString *userId = call.arguments[@"userId"];
        ICMUserAttributes *attributes = [ICMUserAttributes new];
        attributes.userId = userId;
        [Intercom loginUserWithUserAttributes:attributes success:^{
            // Handle success
            result(@"Registered user");
        } failure:^(NSError * _Nonnull error) {
            // Handle failure
            NSInteger errorCode = error.code;
            NSString *errorMsg = error.localizedDescription;
            
            result([FlutterError errorWithCode:[@(errorCode) stringValue]
                                       message:errorMsg
                                       details: [self getIntercomError:errorCode:errorMsg]]);
        }];
    }
    else if([@"loginIdentifiedUserWithEmail" isEqualToString:call.method]) {
        NSString *email = call.arguments[@"email"];
        ICMUserAttributes *attributes = [ICMUserAttributes new];
        attributes.email = email;
        [Intercom loginUserWithUserAttributes:attributes success:^{
            // Handle success
            result(@"Registered user");
        } failure:^(NSError * _Nonnull error) {
            // Handle failure
            NSInteger errorCode = error.code;
            NSString *errorMsg = error.localizedDescription;
            
            result([FlutterError errorWithCode:[@(errorCode) stringValue]
                                       message:errorMsg
                                       details: [self getIntercomError:errorCode:errorMsg]]);
        }];
    }
    else if([@"setLauncherVisibility" isEqualToString:call.method]) {
        NSString *visibility = call.arguments[@"visibility"];
        [Intercom setLauncherVisible:[@"VISIBLE" isEqualToString:visibility]];
        result(@"Setting launcher visibility");
    }
    else if([@"setInAppMessagesVisibility" isEqualToString:call.method]) {
        NSString *visibility = call.arguments[@"visibility"];
        [Intercom setInAppMessagesVisible:[@"VISIBLE" isEqualToString:visibility]];
        result(@"Setting in app messages visibility");
    }
    else if([@"unreadConversationCount" isEqualToString:call.method]) {
        NSUInteger count = [Intercom unreadConversationCount];
        result(@(count));
    }
    else if([@"displayMessenger" isEqualToString:call.method]) {
        [Intercom presentIntercom];
        result(@"Presented messenger");
    }
    else if([@"hideMessenger" isEqualToString:call.method]) {
        [Intercom hideIntercom];
        result(@"Messenger hidden");
    }
    else if([@"displayHelpCenter" isEqualToString:call.method]) {
        [Intercom presentIntercom:helpCenter];
        result(@"Presented help center");
    }
    else if([@"displayHelpCenterCollections" isEqualToString:call.method]) {
        NSArray *collectionIds = call.arguments[@"collectionIds"];
        if(collectionIds != (id)[NSNull null] && collectionIds != nil) {
            [Intercom presentContent:[IntercomContent helpCenterCollectionsWithIds:collectionIds]];
        } else {
            [Intercom presentContent:[IntercomContent helpCenterCollectionsWithIds:@[]]];
        }
        result(@"Presented help center collections");
    }
    else if([@"displayMessages" isEqualToString:call.method]) {
        [Intercom presentIntercom:messages];
        result(@"Presented messages space");
    }
    else if([@"updateUser" isEqualToString:call.method]) {
        [Intercom updateUser:[self getAttributes:call] success:^{
            // Handle success
            result(@"Updated user");
        } failure:^(NSError * _Nonnull error) {
            // Handle failure
            NSInteger errorCode = error.code;
            NSString *errorMsg = error.localizedDescription;
            
            result([FlutterError errorWithCode:[@(errorCode) stringValue]
                                       message:errorMsg
                                       details: [self getIntercomError:errorCode:errorMsg]]);
        }];
    }
    else if([@"logout" isEqualToString:call.method]) {
        [Intercom logout];
        result(@"Logged out");
    }
    else if ([@"logEvent" isEqualToString:call.method]) {
        NSString *name = call.arguments[@"name"];
        NSDictionary *metaData = call.arguments[@"metaData"];
        if(name != (id)[NSNull null] && name != nil) {
            if(metaData != (id)[NSNull null] && metaData != nil) {
                [Intercom logEventWithName:name metaData:metaData];
            } else {
                [Intercom logEventWithName:name];
            }
            result(@"Logged event");
        }
    }
    else if([@"handlePushMessage" isEqualToString:call.method]) {
        result(@"No op");
    }
    else if([@"displayMessageComposer" isEqualToString:call.method]) {
        NSString *message = call.arguments[@"message"];
        [Intercom presentMessageComposer:message];
        result(@"Presented message composer");
    } else if([@"sendTokenToIntercom" isEqualToString:call.method]){
        NSString *token = call.arguments[@"token"];
        if(token != (id)[NSNull null] && token != nil) {
            NSData *encodedToken=[self createDataWithHexString:token];
            // NSData* encodedToken=[token dataUsingEncoding:NSUTF8StringEncoding];
            NSLog(@"%@", encodedToken);
            [Intercom setDeviceToken:encodedToken failure:^(NSError * _Nonnull error) {
                // Handle failure
                NSLog(@"Error setting device token: %@", error.localizedDescription);
            }];
            result(@"Token set");
        }
    } else if([@"displayArticle" isEqualToString:call.method]) {
        NSString *articleId = call.arguments[@"articleId"];
        NSLog(@"%@", articleId);
        if(articleId != (id)[NSNull null] && articleId != nil) {
            [Intercom presentContent:[IntercomContent articleWithId:articleId]];
            result(@"displaying article");
        }
    } else if([@"displayCarousel" isEqualToString:call.method]) {
        NSString *carouselId = call.arguments[@"carouselId"];
        if(carouselId != (id)[NSNull null] && carouselId != nil) {
            [Intercom presentContent:[IntercomContent carouselWithId:carouselId]];
            result(@"displaying carousel");
        }
    } else if([@"displaySurvey" isEqualToString:call.method]) {
        NSString *surveyId = call.arguments[@"surveyId"];
        if(surveyId != (id)[NSNull null] && surveyId != nil) {
            [Intercom presentContent:[IntercomContent surveyWithId:surveyId]];
            result(@"displaying survey");
        }
    } else if([@"isIntercomPush" isEqualToString:call.method]) {
        NSDictionary *message = call.arguments[@"message"];
        if([Intercom isIntercomPushNotification:message]) {
            result(@(YES));
        }else{
            result(@(NO));
        }
    } else if([@"handlePush" isEqualToString:call.method]) {
        NSDictionary *message = call.arguments[@"message"];
        [Intercom handleIntercomPushNotification:message];
        result(@"handle push");
    } else if([@"displayConversation" isEqualToString:call.method]) {
        NSString *conversationId = call.arguments[@"conversationId"];
        if(conversationId != (id)[NSNull null] && conversationId != nil) {
            [Intercom presentContent:[IntercomContent conversationWithId:conversationId]];
            result(@"displaying conversation");
        }
    } else if([@"displayTickets" isEqualToString:call.method]) {
        [Intercom presentIntercom:tickets];
        result(@"Presented tickets space");
    } else if([@"displayHome" isEqualToString:call.method]) {
        [Intercom presentIntercom:home];
        result(@"Presented home space");
    } else if([@"isUserLoggedIn" isEqualToString:call.method]) {
        if([Intercom isUserLoggedIn]) {
            result(@(YES));
        }else{
            result(@(NO));
        }
    } else if([@"fetchLoggedInUserAttributes" isEqualToString:call.method]) {
        ICMUserAttributes *data = [Intercom fetchLoggedInUserAttributes];
        if(data != (id)[NSNull null]){
            NSDictionary *attributes = data.attributes;
            NSMutableDictionary<NSString *, id> *map = [attributes mutableCopy];
            
            // Add custom attributes
            map[@"custom_attributes"] = data.customAttributes;
           
            // Add companies
            if (data.companies) {
                NSMutableArray *companiesArray = [NSMutableArray array];
                for (ICMCompany *company in data.companies) {
                    [companiesArray addObject:[company attributes]];
                }
                map[@"companies"] = companiesArray;
            }
            
            result(map);
        }
        result([NSMutableDictionary dictionary]);
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}

- (NSMutableDictionary *) getIntercomError:(NSInteger)errorCode :(NSString *)errorMessage {
    NSMutableDictionary *details = [NSMutableDictionary dictionary];
    [details setObject:[NSNumber numberWithInteger:errorCode]  forKey: @"errorCode"];
    [details setObject: errorMessage forKey:  @"errorMessage"];
    
    return details;
}

- (ICMUserAttributes *) getAttributes:(FlutterMethodCall *)call {
    ICMUserAttributes *attributes = [ICMUserAttributes new];
    NSString *email = call.arguments[@"email"];
    if(email != (id)[NSNull null]) {
        attributes.email = email;
    }
    NSString *name = call.arguments[@"name"];
    if(name != (id)[NSNull null]) {
        attributes.name = name;
    }
    NSString *phone = call.arguments[@"phone"];
    if(phone != (id)[NSNull null]) {
        attributes.phone = phone;
    }
    NSString *userId = call.arguments[@"userId"];
    if(userId != (id)[NSNull null]) {
        attributes.userId = userId;
    }
    NSString *companyName = call.arguments[@"company"];
    NSString *companyId = call.arguments[@"companyId"];
    if(companyName != (id)[NSNull null] && companyId != (id)[NSNull null]) {
        ICMCompany *company = [ICMCompany new];
        company.name = companyName;
        company.companyId = companyId;
        attributes.companies = @[company];
    }
    NSDictionary *customAttributes = call.arguments[@"customAttributes"];
    if(customAttributes != (id)[NSNull null]) {
        attributes.customAttributes = customAttributes;
    }
    
    NSNumber *signedUpAt = call.arguments[@"signedUpAt"];
    if(signedUpAt != (id)[NSNull null]) {
        attributes.signedUpAt = [NSDate dateWithTimeIntervalSince1970: signedUpAt.doubleValue];
    }
    
    NSString *language = call.arguments[@"language"];
    if(language != (id)[NSNull null]) {
        attributes.languageOverride = language;
    }
    
    return attributes;
}

- (NSData *) createDataWithHexString:(NSString*)inputString {
    NSUInteger inLength = [inputString length];

    unichar *inCharacters = alloca(sizeof(unichar) * inLength);
    [inputString getCharacters:inCharacters range:NSMakeRange(0, inLength)];

    UInt8 *outBytes = malloc(sizeof(UInt8) * ((inLength / 2) + 1));

    NSInteger i, o = 0;
    UInt8 outByte = 0;

    for (i = 0; i < inLength; i++) {
        UInt8 c = inCharacters[i];
        SInt8 value = -1;

        if      (c >= '0' && c <= '9') value =      (c - '0');
        else if (c >= 'A' && c <= 'F') value = 10 + (c - 'A');
        else if (c >= 'a' && c <= 'f') value = 10 + (c - 'a');

        if (value >= 0) {
            if (i % 2 == 1) {
                outBytes[o++] = (outByte << 4) | value;
                outByte = 0;
            } else {
                outByte = value;
            }

        } else {
            if (o != 0) break;
        }
    }

    NSData *a = [[NSData alloc] initWithBytesNoCopy:outBytes length:o freeWhenDone:YES];
    return a;
}
@end
